library/pk.c

```diff
@@ -521,6 +521,76 @@ int mbedtls_pk_sign( mbedtls_pk_context *ctx, mbedtls_md_type_t md_alg,
                                          f_rng, p_rng, NULL ) );
 }
 
+/*
+ * Make a signature with options
+ */
+int mbedtls_pk_sign_ext( mbedtls_pk_type_t type, const void *options,
+             mbedtls_pk_context *ctx, mbedtls_md_type_t md_alg,
+             const unsigned char *hash, size_t hash_len,
+             unsigned char *sig, size_t sig_size, size_t *sig_len,
+             int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
+{
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( ( md_alg == MBEDTLS_MD_NONE && hash_len == 0 ) ||
+                     hash != NULL );
+    PK_VALIDATE_RET( sig != NULL );
+
+    *sig_len = 0;
+    if( ctx->pk_info == NULL )
+        return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+    if( ! mbedtls_pk_can_do( ctx, type ) )
+        return( MBEDTLS_ERR_PK_TYPE_MISMATCH );
+
+    if( type != MBEDTLS_PK_RSASSA_PSS )
+    {
+        /* General case: no options */
+        if( options != NULL )
+            return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+        return( mbedtls_pk_sign_restartable( ctx, md_alg, hash, hash_len,
+                                             sig, sig_size, sig_len,
+                                             f_rng, p_rng, NULL ) );
+    }
+
+#if defined(MBEDTLS_RSA_C) && defined(MBEDTLS_PKCS1_V21)
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    const mbedtls_pk_rsassa_pss_options *pss_opts;
+
+#if SIZE_MAX > UINT_MAX
+    if( md_alg == MBEDTLS_MD_NONE && UINT_MAX < hash_len )
+        return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+#endif /* SIZE_MAX > UINT_MAX */
+
+    if( options == NULL )
+        return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+    pss_opts = (const mbedtls_pk_rsassa_pss_options *) options;
+    if( sig_size < mbedtls_pk_get_len( ctx ) )
+        return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
+
+    if( mbedtls_rsa_set_padding( mbedtls_pk_rsa( *ctx ),
+                                 MBEDTLS_RSA_PKCS_V21,
+                                 md_alg ) != 0 )
+    {
+        return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+    }
+
+    *sig_len = mbedtls_pk_get_len( ctx );
+    ret = mbedtls_rsa_rsassa_pss_sign_ext( mbedtls_pk_rsa( *ctx ),
```

Comment:
This begins with a check of the padding field, which is a problem for us. I can see two ways out of that:

1. Make a copy of this function, for example called `mbedtls_rsa_rsassa_pss_sign_internal()`, declared in `library/rsa_internal.h`that doesn't check the padding field (and make the existing function a wrapper around the new one), then use that `_internal()` function here. It might be the simplest way to do it, but I find it a bit messy because we have several variant of the same function around.
2. Make a temporary copy of the RSA context - then we can set the padding field however we want on this temporary copy. (This is what's done, for different reasons, but also related to thread safety, in `eckey_sign_wrap()`.)

A variant of 2: while we're making a temporary copy, we might as well move the key to PSA and do the operation with PSA. See for example what's done in `mbedtls_pk_verify_ext()` in recent development (that is, now that #5521 has been merged).

I might be biased, but I'm quite inclined to go with the variant of 2: make it PSA-only.

Reply:
I create variant 3. 

create `mbedtls_rsassa_pss_info` in `pk_wrap.c` . 
1. get pk_info_t from pk_type .
2. call sign_func in pk_info. 
Then benefit is we do not need change anything in pk_context .